gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/FlowGraphPathFinder.java

```diff
@@ -0,0 +1,428 @@
+   * @return a list of {@link FlowEdge}s starting at the srcNode and ending at the destNode.
+   */
+  public Dag<JobSpecWithExecutor> findPath() throws PathFinderException {
+    try {
+      //Initialization of auxiliary data structures used for path computation
+      this.pathMap = new HashMap<>();
+      this.flowEdgeToSpecExecutorMap = new HashMap<>();
+      this.flowEdgeToMergedConfig = new HashMap<>();
+
+      // Generate flow execution id for this compilation
+      this.flowExecutionId = System.currentTimeMillis();
+
+      //Path computation must be thread-safe to guarantee read consistency. In other words, we prevent concurrent read/write access to the
+      // flow graph.
+      // TODO: we can easily improve the performance by using a ReentrantReadWriteLock associated with the FlowGraph. This will
+      // allow multiple concurrent readers to not be blocked on each other, as long as there are no writers.
+      synchronized (this.flowGraph) {
+        //Base condition 1: Source Node or Dest Node is inactive; return null
+        if (!srcNode.isActive() || !destNode.isActive()) {
+          log.warn("Either source node {} or destination node {} is inactive; skipping path computation.", srcNodeId,
+              destNodeId);
+          return null;
+        }
+
+        //Base condition 2: Check if we are already at the target. If so, return an empty dag.
+        if ((srcNode.equals(destNode)) && srcDatasetDescriptor.isCompatibleWith(destDatasetDescriptor)) {
+          return new Dag<>(new ArrayList<>());
+        }
+
+        LinkedList<FlowEdgeContext> edgeQueue = new LinkedList<>();
+        edgeQueue.addAll(getNextEdges(srcNode, srcDatasetDescriptor, destDatasetDescriptor, flowConfig));
+        for (FlowEdgeContext flowEdgeContext : edgeQueue) {
+          this.pathMap.put(flowEdgeContext, flowEdgeContext);
+        }
+
+        //At every step, pop an edge E from the edge queue. Mark the edge E as visited. Generate the list of adjacent edges
+        // to the edge E. For each adjacent edge E', do the following:
+        //    1. check if the FlowTemplate described by E' is resolvable using the flowConfig, and
+        //    2. check if the output dataset descriptor of edge E is compatible with the input dataset descriptor of the
+        //       edge E'. If yes, add the edge E' to the edge queue.
+        // If the edge E' satisfies 1 and 2, add it to the edge queue for further consideration.
+        while (!edgeQueue.isEmpty()) {
+          FlowEdgeContext flowEdgeContext = edgeQueue.pop();
+
+          DataNode currentNode = this.flowGraph.getNode(flowEdgeContext.getEdge().getEndPoints().get(1));
+          DatasetDescriptor currentOutputDatasetDescriptor = flowEdgeContext.getOutputDatasetDescriptor();
+
+          //Are we done?
+          if (isPathFound(currentNode, destNode, currentOutputDatasetDescriptor, destDatasetDescriptor)) {
+            return buildDag(flowEdgeContext);
```

Comment:
Why are we putting this in the path finder? Can we move the logic to resolve job specs into a dag from the path into a different class? Path finding and Dag resolution are 2 different responsibilities and should be decoupled. It also reduces the size of this class.

Reply:
Should we consider the multicast case, or like a diamond shape path? For example, a flow might look like A->B->D and A->C->D (or A->B and A->C and A->D). Currently there is no way to specify multiple dst. But is there a way in the future to expand this algorithm to handle multiple cases? Otherwise we will have to create another PathFinder which might have a lot of duplicates. If you need different algorithm, please make this PathFinder as a base class so that more strategy can be applied.

Reply:
@yukuai518 I was planning to have a separate PR to handle the case of multiple destinations. In the first version, single source-multiple destinations will be treated as a sequence of single source single destination, and we will leverage the findPath() for this case.

Reply:
@zxcware Moved the logic of building the job spec out of the class. The rest of the logic is tied to path finding.